Implement .NET external extension system.

darktears opened the pull request:
This commit adds support to load external .NET extensions
for 3rd party developers.

You can specify the path of the extensions to load using
--external-extension-path=/path/to/extension

The extensions are expected to be a .DLL constructed with
Visual Studio as a C# Library (and thus multiarch). Alongside
with that library the 3rd party developer needs to include a .NET bridge
named myextension_bridge.dll given the .NET extension is called
myextension.dll. The dotnet bridge is shipped part of Crosswalk
and is called xwalk_dotnet_bridge and it's a simple copy/rename
for the developer.

This .NET library needs to contains two classes XWalkExtension and
XWalkExtensionInstance which are going to be invoked by Crosswalk
to load the extensions and call the various methods on each instances
to make communication between .NET and JS. .NET extensions are a bit
higher level than our C or C++ extensions because the .NET extension
system will automatically call the constructors in .NET when needed :
e.g. creating the XWalkExtension instance and then create the various
XWalkExtensionInstance instances. This is following inspiration of
what's inside tizen-extension-crosswalk common/ directory which is used
to abstract the low levelness to who write t-e-c extensions.

```
 _______       _______       ______________
| xwalk | <-> | bridge| <-> |.NET extension|
|_______|     |_______|     |______________|
```
The bridge is loaded dynamically like an external extension would
be loaded : it's a DLL and it contains a single exported symbol : XW_Initialize.

One bridge instance is used by each extension and it is responsible to load
the .NET extension, verify using .NET reflection that the entry points
(e.g. XWalkExtension constructor, XWalkInstance handleMessage method...)
are present and that the .NET extension can be used. It then handles the
communication back and forth with Crosswalk and the extension.

One implementation detail worth to note is the usage of the dynamic parameter
in XWalkExtensionInstance, this pointer is used to communicate back from
.NET to C++/XWalk and the fact is dynamic is to hide implementation
details to the 3rd party developer which should only care about calling
the right functions on it. If they don't exists, there will be a runtime
error showing up. It makes using the system very simple (nothing to
include, nothing to import).

Please note that we don't support all the features of the extension
system at the moment, for example you can't add multiple entry points
in JS, and you can't have access to the runtime variables inside the
extension. These features can be added in a following patch.

This commit also roll GYP forward so we can compile passing /clr as
a compiler paramenter. It will probably be not needed by the next
rebase.

Part of the feature work I've added a bunch of tests which
actually build .NET extensions and test the overall system (thus
the .csproj files checked in). There is a little script to invoke
msbuild so it builds the DLL in .NET (GYP doesn't know about that).

darktears commented:
@pozdnyakov  What about this?

darktears commented:
Of course right now this works for one extension in the path. I'm wondering how we could make it work : On way is to have the bridges alongside each extension with the same name for example : fooextension.dll with fooextension_bridge.dll (a copy of xwalk_dotnet_bridge.dll renamed) then barextension.dll with barextension_bridge.dll. Still make the creation of extensions simple and require to copy one file. Better suggestion?

pozdnyakov commented on extensions/common/xwalk_extension_server.cc:
```diff
@@ -406,6 +406,8 @@ std::vector<std::string> RegisterExternalExtensionsInDirectory(
     // load multiple times.
     (*runtime_variables)["extension_path"] =
         new base::StringValue(extension_path.AsUTF8Unsafe());
+    (*runtime_variables)["extension_dir"] =
```
can we get it from "extension_path" when needed?

pozdnyakov commented on extensions/common/xwalk_external_instance.h:
```diff
@@ -30,6 +30,8 @@ class XWalkExternalInstance : public XWalkExtensionInstance {
                         XW_Instance xw_instance);
   ~XWalkExternalInstance() override;
 
+  void* InstanceData() const { return instance_data_; }
```
GetInstanceData also I'd introduce "using InstanceData = void*"

pozdnyakov commented on extensions/public/win/xwalk_dotnet_bridge.cc:
```diff
@@ -0,0 +1,402 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "xwalk/extensions/public/win/xwalk_dotnet_bridge.h"
```
I'd put it into "xwalk/dotnet_extensions" or "xwalk/dotnet'

pozdnyakov commented on extensions/public/win/xwalk_dotnet_bridge.cc:
```diff
@@ -0,0 +1,402 @@
+
+void XWalkDotNetBridge::OnShutdown(XW_Extension) {
+  delete g_bridge;
+  g_bridge = NULL;
```
nit: nullptr

pozdnyakov commented:
I like this approach better than the previous one.

darktears commented:
@pozdnyakov : Implemented what we discussed. Also added a test to verify multiple extensions in the same directory works correctly.

@rakuco : what about the python code?

darktears commented:
In any case please do not merge. I'll do it.

pozdnyakov commented on dotnet/xwalk_dotnet_bridge.cc:
```diff
@@ -0,0 +1,398 @@
+XWalkDotNetBridge::XWalkDotNetBridge()
+  : library_path_(""),
+    initialized_(false),
+    extension_dotnet_(nullptr) {}
```
extension_assembly_(nullptr)

pozdnyakov commented on dotnet/xwalk_dotnet_bridge.cc:
```diff
@@ -0,0 +1,398 @@
+    std::string message;
+    MarshalString(e->ToString(), &message);
+    std::cerr << "Error creating extension instance"
+              << library_path_.c_str() << "': " << message;
```
std::err << std::string should work fine, why c_str()?

darktears replied:
Indeed.

pozdnyakov commented on dotnet/xwalk_dotnet_bridge.cc:
```diff
@@ -0,0 +1,398 @@
+};
+
+XWalkDotNetBridge::XWalkDotNetBridge()
+  : library_path_(""),
```
std::string does not require explicit initialization here

pozdnyakov commented on dotnet/xwalk_dotnet_bridge.cc:
```diff
@@ -0,0 +1,398 @@
+
+public ref class Bridge : public Object {
+public:
+  Bridge(XWalkDotNetBridge* bridge)
```
explicit

darktears replied:
ok

pozdnyakov commented on dotnet/xwalk_dotnet_bridge.cc:
```diff
@@ -0,0 +1,398 @@
+public:
+  Bridge(XWalkDotNetBridge* bridge)
+    : bridge_(bridge),
+    instance_(0) {}
```
instance_(nullptr)

darktears replied:
XW_Instance is defined as an int32_t so I can't do that.

pozdnyakov commented on dotnet/xwalk_dotnet_bridge.h:
```diff
@@ -0,0 +1,45 @@
+  std::string library_path_;
+  bool initialized_;
+  void* extension_dotnet_;
+  void* extension_assembly_;
```
how about struct PrivateData { gcroot<Object^> extension_object; ... extension_assembly; } // in .cc file
and here in class declaration:
PrivateData* p_data_;

darktears replied:
Ok

pozdnyakov commented on dotnet/xwalk_dotnet_bridge.h:
```diff
@@ -0,0 +1,44 @@
+  void SetSyncReply(XW_Instance instance, const std::string& message);
+
+ private:
+   PrivateBridgeData* private_data_;
```
nit: class PrivateBridgeData* private_data_;

pozdnyakov commented on dotnet/xwalk_dotnet_bridge.cc:
```diff
@@ -0,0 +1,414 @@
+  }
+  std::string library_path_;
+  bool initialized_;
+  void* extension_dotnet_;
```
better to use explicit types, no need void*

pozdnyakov commented on dotnet/xwalk_dotnet_bridge.cc:
```diff
@@ -0,0 +1,414 @@
+  std::string library_path_;
+  bool initialized_;
+  void* extension_dotnet_;
+  void* extension_assembly_;
```
ditto

pozdnyakov commented on dotnet/xwalk_dotnet_bridge.cc:
```diff
@@ -0,0 +1,414 @@
+        static_cast<gcroot<Assembly^>*>(private_data_->extension_assembly_);
+    delete extension_assembly_ptr;
+    private_data_->extension_assembly_ = nullptr;
+  }
```
delete private_data_

darktears replied:
oops :)

pozdnyakov commented:
lgtm with comments
